### PR TITLE
[cling] Enable JITLink for AArch64 and x86_64 on Linux

### PR DESCRIPTION
**Pull request by devajithvs**
# This Pull request:

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

### CODE REVIEW

**Comment by vgvassilev** [comment]
That’s a good direction. Let’s wait for the dust to settle after the upgrade for a few weeks before landing this. There were several significant changes between llvm16 and llvm17 in order for jitlink to become default. Can you check the llvm git history?

**Comment by hahnjo** [comment]
> That’s a good direction. Let’s wait for the dust to settle after the upgrade for a few weeks before landing this.

Yes, we discussed to not merge before early next year.

> There were several significant changes between llvm16 and llvm17 in order for jitlink to become default. Can you check the llvm git history?

I put a bit more documentation in https://github.com/root-project/root/issues/14209; the switch to default happened very early in the LLVM 17 cycle (February and April, after branching in January). I was hoping there are only very few fixes around that time, if at all...

**Comment by vgvassilev** [comment]
Hi @lhames, following the deprecation procedure for RuntimeDyld we are hitting a few test failures. Some of them have things like `cling JIT session error: In graph cling-module-315-jitted-objectbuffer, section .eh_frame: relocation target "DW.ref.__gxx_personality_v0" at address 0x7f8d0b32f040 is out of range of Delta32 fixup at 0x7f8c6c00d7c3 (<anonymous block> @ 0x7f8c6c00d7b0 + 0x13)`. You can find more here: https://github.com/root-project/root/runs/28400824352

I am not sure what's our way forward here.

**Comment by hahnjo** [comment]
I now understand what is going wrong: @Axel-Naumann pushed me into the right direction by noticing that the errors concern vtables and typeinfo symbols, which are subject to reemission on our side. Then our `WeakTypeinfoVTablePass` will mark them as `weak` to make sure the JIT infrastructure deduplicates the symbols as needed. Additionally, `ReuseExistingWeakSymbols` will entirely remove the definitions if the symbol can be found in the process, for example for compiled classes. Now the problem is that the symbols were originally marked as "local to the shared library" and relocations can take advantage of this. This is where `PreventLocalOptPass` comes in, but it ran before the two passes mentioned before and therefore doesn't consider these symbols.

So tldr, we can fix the problems by reordering the passes. Let's slot this change in after the upgrade to LLVM 18, which is also where upstream changed to the default to JITLink.

**Review by hahnjo — APPROVED** [review]
Fantastic!

**Comment by hahnjo** [comment]
> Well done. Isn’t that already enabled by default in llvm too?

Yes, already since a while. We couldn't / didn't want to switch because it would have lost `CLING_DEBUG` and `CLING_PROFILE`, which is now available with JITLink after https://github.com/root-project/root/pull/20193
